README.md
serial-button-osc
=================

![red button](media/red_button.jpg)

Send an OSC message when a big red button is pressed.

This code base has been developed by [ZKM | Hertz-Lab](https://zkm.de/en/about-the-zkm/organization/hertz-lab) as part of the project [»The Intelligent Museum«](#the-intelligent-museum). 

Copyright (c) 2023 ZKM | Karlsruhe.  
Copyright (c) 2023 Dan Wilcox.  

BSD Simplified License.

Description
-----------

In ealry 2023, I got a big red button with a USB cable from the ZKM Video Studio. Inside the box is a microcontroller which communicates over a serial port and sends a '3' character when the button is pressed.

This project simply listens for the character and sends an OSC message on a button press. The trigger character and OSC message are configurable. Additionally, the button can act as a switch and send the on.off value on a change aka act as a switch.

Dependencies
------------

* Python 3
* [pyserial](https://github.com/pyserial/pyserial)
* [python-osc](https://github.com/attwad/python-osc)

Setup
-----

Install Python 3, if not already available. For instance, on macOS using [Homebrew](http://brew.sh):

```shell
brew install python3
```

Create a virtual environment and install the script's dependencies:

```shell
make
```

Running
-------

Next, start listening on a serial port via the virtual environment wrapper script:

    ./serial-button-osc /dev/tty.usbserial-310

The program will simply run in a loop and sends a `/button` OSC message whenever a '3' character is received. To modify behavior, use the commandline option flags.

`./serial-button-osc` help output:
~~~
usage: serial-button-osc.py [-h] [-a ADDR] [-p PORT] [-r RATE] [--button-char BUTTON_CHAR] [-s] [--switch-chars SWITCH_CHARS] [-v] [DEV] [MESSAGE]

send an OSC message when a big red button is pressed

positional arguments:
  DEV                   serial port device, default: /dev/ttyAMA0
  MESSAGE               OSC message address, default: /button

options:
  -h, --help            show this help message and exit
  -a ADDR, --address ADDR
                        destination hostname or IP address, default: 127.0.0.1
  -p PORT, --port PORT  destination port to send to, default: 6000
  -r RATE, --rate RATE  serial port baud rate, default: 115200
  --button-char BUTTON_CHAR
                        serial char for button press, default: '3'
  -s, --switch          read button as a switch and send off/on int value as message arg
  --switch-chars SWITCH_CHARS
                        serial chars for switch values off/on, default: '01'
  -v, --verbose         enable verbose printing
~~~

For example, to send "/do/something" when a 'b' char is received over the "/dev/tty.usbserial-310" serial device:

    ./serial-button-osc.py -v --button-char b /dev/tty.usbserial-310 /do/something

Additionally, to have the button act as a switch and send "/record f" when either 'A' or 'B' are received:

    ./serial-button-osc.py -v --switch-chars AB /dev/tty.usbserial-310 /record

The following messages will be sent based on the character read:
* 'A': `/record 0`
* 'B': `/record 1`

To stop serial-button-osc, use CTRL+C to issue an interrupt signal.

### Calling Python script directly

The Python script can be called directly without the wrapper script, but requires manually enabling or disabling the virtual environment:

Activate the virtual environment before the first run in a new commandline session:

    source venv/bin/activate

Use:

    ./baton.py -h

When finished, deactivate the virtual environment with:

    deactivate

Example Clients
---------------

An example clients is included:

* pd/recvclient.pd: Pure Data patch which receives OSC messages

Start the recv client and serial-button-osc. Pressing the button should result in a message receive event and a virtual button flash in the pd patch.

Making a Button/Switch with an Arduino
--------------------------------------

![style button](media/style-button.jpg)

![style button inside](media/style-button-inside.jpg)

_Example: [Styler](https://github.com/zkmkarlsruhe/Styler) style input trigger button._

A simple serial button can be made using an analog button or switch with an Arduino microcontroller.

Basic wiring diagram:
~~~
3.3V -> button -> 10k Ohm -> GND
              \-> D2
~~~

To create a serial button:
1. Connect a button to an Arduino using the above diagram
2. Plug the Arduino into the development computer
3. Using the Arduino software, compile and upload the `arduino/ButtonChar` sketch to the Arduino
4. To test, enable the Arduino serial console, select a baud of 115200, and look for a char when the button is pressed

To have the button act as a switch, use the same wiring diagram with the `arduino/SwitchChar` sketch which by default sends a `1` when the button is pressed and a `0` on release.

Checking Serial Connection
--------------------------

On macOS and Linux, the serial port can be connected to as a text console via `cu`:

    sudo cu -s 115200 -l /dev/tty.usbserial-310

When connected, cu will print received bytes as plain ASCII text.

To exit, type "~." and wait for cu to disconnect.

The Intelligent Museum
----------------------

An artistic-curatorial field of experimentation for deep learning and visitor participation

The [ZKM | Center for Art and Media](https://zkm.de/en) and the [Deutsches Museum Nuremberg](https://www.deutsches-museum.de/en/nuernberg/information/) cooperate with the goal of implementing an AI-supported exhibition. Together with researchers and international artists, new AI-based works of art will be realized during the next four years (2020-2023).  They will be embedded in the AI-supported exhibition in both houses. The Project „The Intelligent Museum” is funded by the Digital Culture Programme of the [Kulturstiftung des Bundes](https://www.kulturstiftung-des-bundes.de/en) (German Federal Cultural Foundation) and funded by the [Beauftragte der Bundesregierung für Kultur und Medien](https://www.bundesregierung.de/breg-de/bundesregierung/staatsministerin-fuer-kultur-und-medien) (Federal Government Commissioner for Culture and the Media).

As part of the project, digital curating will be critically examined using various approaches of digital art. Experimenting with new digital aesthetics and forms of expression enables new museum experiences and thus new ways of museum communication and visitor participation. The museum is transformed to a place of experience and critical exchange.

![Logo](media/Logo_ZKM_DMN_KSB.png)


CHANGES.txt
0.2.0: 2023 Feb 02

* only print button pressed if verbose

* can now read button as a switch with off/on events
* added -s/--switch mode and --switch-chars
* added arudino sketch and info on creating a serial button or switch
* added note about using cu to read serial port on commandline
* added icon

* renamed -c/--char to --button-char

0.1.0: 2023 Jan 05

initial version


serial-button-osc.py
#! /usr/bin/env python3
#
# Copyright (c) 2023 ZKM | Hertz-Lab
# Dan Wilcox <dan.wilcox@zkm.de>
#
# BSD Simplified License.
# For information on usage and redistribution, and for a DISCLAIMER OF ALL
# WARRANTIES, see the file, "LICENSE.txt," in this distribution.
#
# This code has been developed at ZKM | Hertz-Lab as part of „The Intelligent
# Museum“ generously funded by the German Federal Cultural Foundation.

import sys
import signal
import argparse

##### parser

def len_min_2(s):
    if len(s) > 1:
        return s
    raise argparse.ArgumentTypeError("value must have min len of 2")

parser = argparse.ArgumentParser(description="send an OSC message when a big red button is pressed")
parser.add_argument(
    "dev", type=str, nargs="?", metavar="DEV",
    default="/dev/ttyAMA0", help="serial port device, default: /dev/ttyAMA0")
parser.add_argument(
    "message", type=str, nargs="?", metavar="MESSAGE",
    default="/button", help="OSC message address, default: /button")
parser.add_argument(
    "-a", "--address", dest="address", metavar="ADDR",
    default="127.0.0.1", help="destination hostname or IP address, default: 127.0.0.1")
parser.add_argument(
    "-p", "--port", type=int, dest="port", metavar="PORT",
    default=6000, help="destination port to send to, default: 6000")
parser.add_argument(
    "-r", "--rate", type=int, dest="rate", metavar="RATE",
    default=115200, help="serial port baud rate, default: 115200")
parser.add_argument(
    "--button-char", dest="buttonchar", metavar="BUTTON_CHAR",
    default="3", help="serial char for button press, default: '3'")
parser.add_argument(
    "-s", "--switch", action="store_true", dest="switch",
    help="read button as a switch and send off/on int value as message arg")
parser.add_argument(
    "--switch-chars", dest="switchchars", metavar="SWITCH_CHARS",
    default="01", help="serial chars for switch values off/on, default: '01'",
    type=len_min_2) # restrict to min len of 2
args = parser.add_argument(
    "-v", "--verbose", action="store_true", dest="verbose",
    help="enable verbose printing")

##### Serial

import serial
import time

# serial byte reader base class
class SerialReader:

    def __init__(self, dev, rate=115200, verbose=False):
        self.serial = serial.Serial(dev, rate) # serial port
        self.interval = 0.1
        self.is_running = True
        self.verbose = verbose
        if self.verbose:
            print(f"serial: created {dev} {rate}")

    # open serial port for reading
    def open(self):
        if not self.serial.is_open:
            self.serial.open()
            if self.verbose:
                print("serial: open")

    # close serial port
    def close(self):
        if self.serial.is_open:
            self.serial.close()
            if self.verbose:
                print("serial: close")

    # start synchronous run loop
    def start(self):
        self.is_running = True
        if self.verbose:
            print("serial: start")
        while self.is_running:
            self.update()
            time.sleep(self.interval)
        self.is_running = False
        if self.verbose:
            print("serial: stop")

    # stop synchronous run loop
    def stop(self):
        self.is_running = False

    # read serial bytes, send message if self.char byte is found
    def update(self):
        count = self.serial.in_waiting
        if count > 0:
            recv = self.serial.read(count)
            self.serial.reset_input_buffer()
            if self.verbose:
                print(f"serial: {recv}")
            self.recv(recv)

    # process received bytes, implement in child class
    def recv(serial_bytes):
        pass

# single char button press
class SerialButton(SerialReader):

    def __init__(self, dev, char, rate, verbose):
        super().__init__(dev, rate, verbose)
        self.char = ord(char) # button press event char byte value
        self.on_press = None  # button press callback: def button_pressed(button)

    # read serial bytes, send message if self.char byte is found
    def recv(self, serial_bytes):
        for c in serial_bytes: # send once per message
            if c == self.char:
                if self.verbose:
                    print(f"button: pressed")
                if self.on_press: self.on_press(self)
                break

# switch value char switch
class SerialSwitch(SerialReader):

    def __init__(self, dev, chars, rate, verbose):
        super().__init__(dev, rate, verbose)
        self.chars = [ord(chars[0]), ord(chars[1])] # switch value event char byte values: [off, on]
        self.on_change = None # switch value change callback: def switch_changed(switch, value)

    # read serial bytes, send message if either byte in self.chars is found
    def recv(self, serial_bytes):
        for c in serial_bytes:
            value = -1
            if c == self.chars[0]:
                value = 0 # off
            elif c == self.chars[1]:
                value = 1 # on
            if value != -1:
                if self.verbose:
                    print(f"switch: {value}")
                if self.on_change: self.on_change(self, value)

##### button/switch -> osc

from pythonosc.udp_client import SimpleUDPClient
from pythonosc import osc_message_builder

# send message on button press event
def button_pressed(button):
    message = osc_message_builder.OscMessageBuilder(address=args.message)
    sender.send(message.build())

# send message on switch value change event
def switch_changed(switch, value):
    message = osc_message_builder.OscMessageBuilder(address=args.message)
    message.add_arg(value)
    sender.send(message.build())

##### signal

# signal handler for nice exit
def sigint_handler(signum, frame):
    reader.stop()

##### main

if __name__ == '__main__':

    # parse
    args = parser.parse_args()

    # button/switch
    try:
        if args.switch:
            reader = SerialSwitch(dev=args.dev, chars=args.switchchars, rate=args.rate,
                                  verbose=args.verbose)
        else:
            reader = SerialButton(dev=args.dev, char=args.buttonchar, rate=args.rate,
                                  verbose=args.verbose)
    except Exception as e:
        print(e)
        exit(1)

    # osc sender
    sender = SimpleUDPClient(args.address, args.port)
    if args.switch:
        reader.on_change = switch_changed
    else:
        reader.on_press = button_pressed

    # start
    signal.signal(signal.SIGINT, sigint_handler)
    try:
        reader.open()
        reader.start()
    finally:
        reader.close()
